core/src/main/java/org/apache/iceberg/expressions/ExpressionParser.java

```diff
@@ -0,0 +1,225 @@
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Set;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.util.JsonUtil;
+
+public class ExpressionParser {
+
+  private static final String TYPE = "type";
+  private static final String VALUE = "value";
+  private static final String OPERATION = "operation";
+  private static final String LITERALS = "literals";
+  private static final String TERM = "term";
+  private static final String LEFT_OPERAND = "left-operand";
+  private static final String RIGHT_OPERAND = "right-operand";
+  private static final String OPERAND = "operand";
+  private static final String AND = "and";
+  private static final String OR = "or";
+  private static final String NOT = "not";
+  private static final String TRUE = "true";
+  private static final String FALSE = "false";
+  private static final String UNBOUNDED_PREDICATE = "unbounded-predicate";
+  private static final String BOUNDED_LITERAL_PREDICATE = "bounded-literal-predicate";
+  private static final String BOUNDED_SET_PREDICATE = "bounded-set-predicate";
+  private static final String BOUNDED_UNARY_PREDICATE = "bounded-unary-predicate";
+  private static final String NAMED_REFERENCE = "named-reference";
+  private static final String BOUND_REFERENCE = "bound-reference";
+  private static final String ABOVE_MAX = "above-max";
+  private static final String BELOW_MIN = "below-min";
+
+  private static final Set<Expression.Operation> oneInputs = ImmutableSet.of(
+          Expression.Operation.IS_NULL,
+          Expression.Operation.NOT_NULL,
+          Expression.Operation.IS_NAN,
+          Expression.Operation.NOT_NAN);
+
+
+  private ExpressionParser() {
+  }
+
+  public static String toJson(Expression expression, boolean pretty) {
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      if (pretty) {
+        generator.useDefaultPrettyPrinter();
+      }
+      toJson(expression, generator);
+      generator.flush();
+      return writer.toString();
+
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to write json", e);
+    }
+  }
+
+  public static void toJson(Expression expression, JsonGenerator generator) throws IOException {
+    if (expression instanceof And) {
+      toJson((And) expression, generator);
+    } else if (expression instanceof Or) {
+      toJson((Or) expression, generator);
+    } else if (expression instanceof Not) {
+      toJson((Not) expression, generator);
+    } else if (expression instanceof True) {
+      toJson((True) expression, generator);
+    } else if (expression instanceof False) {
+      toJson((False) expression, generator);
+    } else if (expression instanceof Predicate) {
+      toJson((Predicate) expression, generator);
+    } else {
+      throw new IllegalArgumentException("Invalid Operation Type");
```

Comment:
Nit: Can you log the received unexpected type?

Also, this covers every `Expresson` subclass I can think of. Are there other `Expression` classes? Maybe `throw new IllegalArgumentException("Unknown Expression type: %s", expression)`